### PR TITLE
Rustup to rust-lang/rust#64874

### PR DESCRIPTION
**Pull request by flip1995**
TODO:
- [x] replace `rvalue_promotable_map` in [1]
- [ ] ~~fix [2] according to this comment https://github.com/rust-lang/rust/pull/64874#issuecomment-536203626 this should be merged with `consume`, but I didn't figure out how to merge them, yet.~~
- [ ] ~~fix [3]; What to do with `LoanCause`?~~

[2]+[3] probably have to be resolved by a rewrite of the lint. https://github.com/rust-lang/rust-clippy/pull/4628#issuecomment-538574944

[1]
https://github.com/rust-lang/rust-clippy/blob/54bf4ffd626970e831bb80c037f804a3b3450835/clippy_lints/src/methods/mod.rs#L1292-L1299

[2]
https://github.com/rust-lang/rust-clippy/blob/54bf4ffd626970e831bb80c037f804a3b3450835/clippy_lints/src/escape.rs#L126

[3]
https://github.com/rust-lang/rust-clippy/blob/54bf4ffd626970e831bb80c037f804a3b3450835/clippy_lints/src/escape.rs#L166-L176

I could need some help with [1]. The purpose of this is to "don't lint for constant values". cc @matthewjasper 

For now I see what I can do with [2].

changelog: Temporary break `boxed_local` lint.

### CODE REVIEW

File: clippy_lints/src/needless_pass_by_value.rs [hunk 1]
```diff
@@ -325,115 +324,39 @@ fn requires_exact_signature(attrs: &[Attribute]) -> bool {
...
+impl MovedVariablesCtxt {
+    fn move_common(&mut self, cmt: &mc::cmt_<'_>) {
         let cmt = unwrap_downcast_or_interior(cmt);
 
         if let mc::Categorization::Local(vid) = cmt.cat {
             self.moved_vars.insert(vid);
         }
     }
-
-    fn non_moving_pat(&mut self, matched_pat: &Pat, cmt: &mc::cmt_<'tcx>) {
```
Comment (flip1995) [inline clippy_lints/src/needless_pass_by_value.rs]: Is removing this completely ok? It didn't break anything in the tests :thinking:

Comment (sinkuu) [reply clippy_lints/src/needless_pass_by_value.rs]: Nothing is inserted to `spans_need_deref` after this change, so [the suggestions to add `*`](https://github.com/rust-lang/rust-clippy/blob/master/tests/ui/needless_pass_by_value.stderr#L166) should be broken, but idk.

Anyway, Rust has gained default binding mode and this suggestion became useless. I think it's OK to remove it.

**Comment by lzutao** [comment]
matthew left some guides here: https://github.com/rust-lang/rust/pull/64874#issuecomment-536203626

**Comment by flip1995** [comment]
Update: I won't get to this until tomorrow. It would be great if someone could take over :)

**Comment by matthewjasper** [comment]
[1] Change `is_ctor_function` to check for promotable functions as well (and give it a different name). It would look something like

```rust
// ...
match cx.tables.qpath_res(qp, fun.hir_id) {
    def::Res::Def(DefKind::Variant, ..) | Res::Def(DefKind::Ctor(..), _) => true,
    // Maybe consider `is_const_fn`, but there's less guarantee of cheapness there.
    def::Ref::Def(DefKind::Fn, def_id) => cx.tcx.is_promotable_const_fn(def_id),
    _ => false,
}
```

[2] and [3] Looking into this lint some more, it's hopelessly broken and needs a rewrite. That said, to roughly keep the current behavior:

[2]:
* Move the ` if is_argument` part into `consume`, using `cmt.hir_id` instead of `consume_pat.hir_id`.
* Remove the `if let Categorization::Rvalue(..) = ...` block, it doesn't appear to work. 
* Move the `if let Categorization::Local(lid) = ...` block into `consume` adding a check that checking that `map.find(cmt.hir_id)` is `Some(Node::Binding(_))` 

[3]: Unconditionally call `self.set.remove(&lid);` - this slightly changes the behavior of this lint around closures.

**Comment by flip1995** [comment]
Thanks a lot for the hints @matthewjasper!

[1] That doesn't work, because the `PathSegment` of a const fn call like `u32::max_value()` always has  `res: Some(Err)`, e.g.
```rust
let _ = 1u32.checked_add(1).unwrap_or(u32::max_value());
```
`dbg!(qpath)`:
```
[clippy_lints/src/utils/mod.rs:809] qp = TypeRelative(
    type(u32),
    PathSegment {
        ident: max_value#0,
        hir_id: Some(
            HirId {
                owner: DefIndex(17),
                local_id: 43,
            },
        ),
        res: Some(
            Err,
        ),
        args: None,
        infer_args: true,
    },
)
```

What does `Res::Err` mean? That the resolution errored or that no resolution is known at this point? (Does `Res` even stand for Resolution?) Sorry if that's a stupid question, but `def::Res` is completely undocumented and I couldn't find anything in the rustc-guide. 

---

> it's hopelessly broken and needs a rewrite

That's what I was afraid of. I propably won't have the time of doing a rewrite, so I'd suggest to just adapt the tests, since this only introduces FNs. We then can get a working Clippy faster and work on this later. cc @rust-lang/clippy 

[2]+[3] I did this is in flip1995/rust-clippy@b7d473503b051f86c59f3256130e8ff13afe4332, but the result is, that all tests fail for this lint

**Comment by flip1995** [comment]
On another note: I think it's time to [Nuke the clippy plugin interface](https://github.com/rust-lang/rust-clippy/pull/2783)

```
warning: use of deprecated attribute `plugin_registrar`: compiler plugins are deprecated. See https://github.com/rust-lang/rust/issues/29597
  --> src/lib.rs:12:1
   |
12 | #[plugin_registrar]
   | ^^^^^^^^^^^^^^^^^^^ help: remove this attribute
   |
   = note: `#[warn(deprecated)]` on by default
```
https://travis-ci.com/rust-lang/rust-clippy/jobs/242398220#L890-L896

**Comment by flip1995** [comment]
As suggested in https://github.com/rust-lang/rust-clippy/pull/4616#issuecomment-538457618

@bors treeclosed=10 p=10

**Comment by matthewjasper** [comment]
> What does `Res::Err` mean? That the resolution errored or that no resolution is known at this point? (Does `Res` even stand for Resolution?) Sorry if that's a stupid question, but `def::Res` is completely undocumented and I couldn't find anything in the rustc-guide.

`Res::Err` usually means that name resolution errored, but it's also used here because the name won't be resolved until type checking is done.

> [2]+[3] I did this is in b7d4735, but the result is, that all tests fail for this lint

I've had a look at this, moving the `is_argument` check to `mutate` fixes this (only the error that I expected to go is now gone).

File: tests/ui/escape_analysis.stderr [hunk 2]
```diff
@@ -12,11 +12,5 @@ error: local variable doesn't need to be boxed here
 LL | pub fn new(_needs_name: Box<PeekableSeekable<&()>>) -> () {}
    |            ^^^^^^^^^^^
 
-error: local variable doesn't need to be boxed here
```
Comment (flip1995) [inline tests/ui/escape_analysis.stderr]: This is the only remaining FN introduced in this PR now. So ping @rust-lang/clippy for review.

Comment (Manishearth) [reply tests/ui/escape_analysis.stderr]: file an issue for it and refer to it in the test file and we can land tis

**Pull request merged**